File: GreatCircle.rb
class GreatCircle
  def self.distance(lat1, lon1, lat2, lon2, unit='KM')
    radius = get_radius unit

    Math::acos(
      Math::sin(to_radians(lat1))*Math::sin(to_radians(lat2)) +
      Math::cos(to_radians(lat1))*Math::cos(to_radians(lat2)) *
      Math::cos(to_radians(lon2-lon1))
    ) * radius
  end

  def self.bearing(lat1, lon1, lat2, lon2)
    lat1 = to_radians lat1
    lat2 = to_radians lat2
    dLon = to_radians lon2-lon1

    y = Math::sin(dLon) * Math::cos(lat2)
    x = Math::cos(lat1) * Math::sin(lat2) -
        Math::sin(lat1) * Math::cos(lat2) * Math::cos(dLon)

    to_bearing Math::atan2(y, x)
  end

  def self.destination(lat1, lon1, brng, distance, unit='KM')
    radius = get_radius unit
    lat1 = to_radians lat1
    lon1 = to_radians lon1
    brng = to_radians brng

    lat2 = Math::asin(
            Math::sin(lat1) * Math::cos(distance/radius) +
            Math::cos(lat1) * Math::sin(distance/radius) *
            Math::cos(brng)
          )

    lon2 = lon1 + Math::atan2(
                    Math::sin(brng) * Math::sin(distance/radius) * Math::cos(lat1),
                    Math::cos(distance/radius) - Math::sin(lat1) * Math::sin(lat2)
                  )

    lon2 = (lon2 + Math::PI) % (2 * Math::PI) - Math::PI

    { 'LAT' => to_degrees(lat2), 'LON' => to_degrees(lon2) }
  end

  private

  # degrees to radians
  def self.to_radians(degrees)
    degrees * Math::PI / 180.0
  end

  # radians to bearing
  def self.to_bearing(radians)
    (to_degrees(radians) + 360.0) % 360.0
  end

  # radians to degree
  def self.to_degrees(radians)
    radians * 180 / Math::PI
  end

  # radius of earth
  def self.get_radius(unit)
    radius = { 'M' => 6371009, 'KM' => 6371.009, 'MI' => 3958.761, 'NM' => 3440.070, 'YD' => 6967420, 'FT' => 20902260 }

    radius[unit] || unit
  end
end

File: Go/GreatCircle/GreatCircle.go
package GreatCircle

import (
	"math"
)

var FixedValues = map[string]float64{
	"M":  6371009,
	"KM": 6371.009,
	"MI": 3958.761,
	"NM": 3440.070,
	"YD": 6967420,
	"FT": 20902260,
}

type Unit interface {
	validateRadius() float64
}

type UnitFixed string
type UnitValue float64

func (r UnitFixed) validateRadius() float64 {
	return FixedValues[string(r)]
}
func (r UnitValue) validateRadius() float64 {
	return float64(r)
}

func deg2rad(deg float64) float64 {
	return deg * math.Pi / 180
}
func rad2deg(rad float64) float64 {
	return rad * 180 / math.Pi
}

func Distance(lat1 float64, lon1 float64, lat2 float64, lon2 float64, unit Unit) float64 {
	r := unit.validateRadius()
	lat1 = deg2rad(lat1)
	lon1 = deg2rad(lon1)
	lat2 = deg2rad(lat2)
	lon2 = deg2rad(lon2)
	lonDelta := lon2 - lon1
	a := math.Pow(math.Cos(lat2)*math.Sin(lonDelta), 2) + math.Pow(math.Cos(lat1)*math.Sin(lat2)-math.Sin(lat1)*math.Cos(lat2)*math.Cos(lonDelta), 2)
	b := math.Sin(lat1)*math.Sin(lat2) + math.Cos(lat1)*math.Cos(lat2)*math.Cos(lonDelta)
	angle := math.Atan2(math.Sqrt(a), b)

	return angle * r
}

func Bearing(lat1 float64, lon1 float64, lat2 float64, lon2 float64) float64 {
	lat1 = deg2rad(lat1)
	lon1 = deg2rad(lon1)
	lat2 = deg2rad(lat2)
	lon2 = deg2rad(lon2)
	lonDelta := lon2 - lon1
	y := math.Sin(lonDelta) * math.Cos(lat2)
	x := math.Cos(lat1)*math.Sin(lat2) - math.Sin(lat1)*math.Cos(lat2)*math.Cos(lonDelta)
	brng := math.Atan2(y, x)
	brng = brng * (180 / math.Pi)
	if brng < 0 {
		brng += 360
	}

	return brng
}

func Destination(lat1 float64, lon1 float64, brng float64, dt float64, unit Unit) map[string]float64 {
	r := unit.validateRadius()
	lat1 = deg2rad(lat1)
	lon1 = deg2rad(lon1)
	lat3 := math.Asin(math.Sin(lat1)*math.Cos(dt/r) + math.Cos(lat1)*math.Sin(dt/r)*math.Cos(deg2rad(brng)))
	lon3 := lon1 + math.Atan2(math.Sin(deg2rad(brng))*math.Sin(dt/r)*math.Cos(lat1), math.Cos(dt/r)-math.Sin(lat1)*math.Sin(lat3))

	return map[string]float64{
		"LAT": rad2deg(lat3),
		"LON": rad2deg(lon3),
	}
}
